android/java/org/chromium/chrome/browser/app/BraveActivity.java

```diff
@@ -720,12 +725,44 @@ protected void initializeStartupMetrics() {
         BraveHelper.DisableFREDRP();
     }
 
+    @Override
+    public void onPreferenceChange() {
+        solveAdaptiveCaptcha();
+    }
+
+    public void solveAdaptiveCaptcha() {
+        String captchaID = UserPrefs.get(Profile.getLastUsedRegularProfile())
+                                   .getString(BravePref.SCHEDULED_CAPTCHA_ID);
+        String paymentID = UserPrefs.get(Profile.getLastUsedRegularProfile())
+                                   .getString(BravePref.SCHEDULED_CAPTCHA_PAYMENT_ID);
+        if (!TextUtils.isEmpty(captchaID) && !TextUtils.isEmpty(paymentID)
+                && !BravePrefServiceBridge.getInstance().getSafetynetCheckFailed()) {
+            BraveAdaptiveCaptchaUtils.solveCaptcha(captchaID, paymentID);
+        }
+    }
+
     @Override
     public void finishNativeInitialization() {
         super.finishNativeInitialization();
 
         BraveHelper.maybeMigrateSettings();
 
+        PrefChangeRegistrar mPrefChangeRegistrar = new PrefChangeRegistrar();
+        mPrefChangeRegistrar.addObserver(BravePref.SCHEDULED_CAPTCHA_ID, this);
+
+        if (UserPrefs.get(Profile.getLastUsedRegularProfile())
+                        .getInteger(BravePref.SCHEDULED_CAPTCHA_FAILED_ATTEMPTS)
+                == 10) {
+            UserPrefs.get(Profile.getLastUsedRegularProfile())
+                    .setBoolean(BravePref.SCHEDULED_CAPTCHA_PAUSED, true);
+        }
+
+        Log.e("adaptive captcha", "Failed attempts : "+ UserPrefs.get(Profile.getLastUsedRegularProfile()).getInteger(BravePref.SCHEDULED_CAPTCHA_FAILED_ATTEMPTS));
```

Comment:
do we need that log?

Reply:
@SergeyZhukovsky it's mainly for QA team to verify as the solution for the captcha happens in the background without any UI changes.